LanguageParser.hs

```diff
@@ -1,40 +1,414 @@
+import qualified Data.Set             as Set
+import           Data.Tuple
+import           Data.Void
+import qualified Text.Megaparsec      as P
+import qualified Text.Megaparsec.Char as P
+import           Text.Numeral.Roman
+import qualified WordLists            as W
 
-import AST
+type Parser = P.Parsec Void String
 
-programP :: P.Parser Program
-programP = undefined
+-- TODO: How could this file be organized better to be more readable?
 
-headerP :: P.Parser Header
-headerP = undefined
+-- Utility functions --
 
-actP :: P.Parser (Label, Act)
-actP = undefined
+tillNoInclC :: Char -> Parser String
+tillNoInclC c = P.many (P.noneOf [c]) <* P.string [c]
 
-sceneP :: P.Parser (Label, Scene)
-sceneP = undefined
+isEndPunctuation :: Char -> Bool
+isEndPunctuation c = (c == '.') || (c == '?') || (c == '!')
 
-statementP :: P.Parser (Statement, Annotation)
-statementP = undefined
+parseUntilEndPunc :: Parser String
+parseUntilEndPunc = P.takeWhileP Nothing (not . isEndPunctuation)
 
-expressionP :: P.Parser Expression
-expressionP = undefined
+-- like parseUntilEndPunc but also parses the end punctuation
+parseUntilEndPunc' :: Parser String
+parseUntilEndPunc' = liftA2 (++)
+                     parseUntilEndPunc (P.takeWhileP Nothing isEndPunctuation)
 
-valueP :: P.Parser Value
-valueP = undefined
+parseUntilEndBracket :: Parser String
+parseUntilEndBracket = P.takeWhileP Nothing (/= ']')
 
-relationshipP :: P.Parser Relationship
-relationshipP = undefined
+-- parse one of the given strings, case-insensitively
+oneOfString' :: [String] -> Parser String
+oneOfString' l = P.choice ((\s -> P.try (P.string' s <*
+                 P.notFollowedBy invalid)) <$> l)
+                 where
+                   -- a word can only be followed by either a space or end punctuation
+                   invalid :: Parser Char
+                   invalid = P.satisfy (\x -> isAlpha x || x == '-')
 
-comparisonP :: P.Parser Comparison
-comparisonP = undefined
+constP :: a -> Parser b -> Parser a
+constP a p = const a <$> (p <* P.space)
 
-sentenceP :: P.Parser Sentence
-sentenceP = undefined
+oneOfCharacterNames :: Parser CName
+oneOfCharacterNames = map toLower <$>
+                      oneOfString' W.characters
+                      P.<?>
+                      "a valid Shakespeare character"
+
+oneOfSecondPersonPos :: Parser String
+oneOfSecondPersonPos = oneOfString' W.secondPersonPossessive
+                       P.<?> "a second person possesive"
+
+-- TODO: Next are compile time checks.
+testParse file = do
+                  s <- readFile file
+                  case P.parse programP file s of
+                    Left err -> do
+                      putStrLn (P.parseErrorPretty' s err)
+                      return ()
+                    Right p -> do
+                      print p
+                      return ()
+
+programP :: Parser Program
+programP = liftA2 Program headerP (Map.fromList <$> many actP) <* P.eof
```

Comment:
I mean. This works, so it's fine if you prefer it.

I prefer fmap/zap because it seems more idiomatic and reads more clearly to me. There's a more clear feeling separation between the arguments (your constructor is clearly separated from the args by `<$>` and the args from each other with explicit `<*>`) which I like.

No big deal, it just reads more nicely to me.